File: go/schemify/internal/schema/drift.go
package schema

import (
	"fmt"
	"regexp"
	"strings"
)

func stripCommentPrefix(line string) string {
	before, after, found := strings.Cut(line, "--")
	if !found {
		return line
	}
	return before + strings.TrimPrefix(after, " ")
}

func netParenDepth(line string) int {
	depth := 0
	inSingle := false
	inDouble := false
	for _, ch := range line {
		switch ch {
		case '\'':
			if !inDouble {
				inSingle = !inSingle
			}
		case '"':
			if !inSingle {
				inDouble = !inDouble
			}
		case '(':
			if !inSingle && !inDouble {
				depth++
			}
		case ')':
			if !inSingle && !inDouble {
				depth--
			}
		}
	}
	return depth
}

var driftOpenRE = regexp.MustCompile(`(?i)^\s*--\s*DRIFT\s+(\w+)\s+(DROP|DEPRECATED)\s*\(\s*$`)
var driftCloseRE = regexp.MustCompile(`^\s*--\s*\)\s*;?\s*$`)

func extractDriftBlocks(rawSQL, tableSchema, tableName string) ([]DriftBlock, error) {
	lines := strings.Split(rawSQL, "\n")
	var blocks []DriftBlock
	seenIDs := make(map[string]DriftPolicy)

	scope := DriftScopeFile
	if tableSchema != "" || tableName != "" {
		scope = DriftScopeTable
	}

	inBlock := false
	var currentID string
	var currentPolicy DriftPolicy
	var bodyLines []string
	depth := 0

	for _, line := range lines {
		if !inBlock {
			m := driftOpenRE.FindStringSubmatch(line)
			if m == nil {
				continue
			}
			id := m[1]
			policy := DriftPolicy(strings.ToUpper(m[2]))
			if existingPolicy, ok := seenIDs[id]; ok && existingPolicy != policy {
				return nil, fmt.Errorf("DRIFT block %q has conflicting policies: %s and %s", id, existingPolicy, policy)
			}
			seenIDs[id] = policy
			inBlock = true
			currentID = id
			currentPolicy = policy
			bodyLines = nil
			depth = 1
		} else {
			strippedLine := stripCommentPrefix(line)
			isClosing := driftCloseRE.MatchString(line)
			net := netParenDepth(strippedLine)
			newDepth := depth + net
			if newDepth == 0 {
				if isClosing {
					blocks = append(blocks, DriftBlock{
						ID:          currentID,
						Policy:      currentPolicy,
						RawBody:     strings.Join(bodyLines, "\n"),
						Scope:       scope,
						TableSchema: tableSchema,
						TableName:   tableName,
					})
					inBlock = false
					continue
				}
				return nil, fmt.Errorf("DRIFT block %q: depth reached 0 before closing -- )", currentID)
			}
			if newDepth < 0 {
				return nil, fmt.Errorf("DRIFT block %q: negative paren depth", currentID)
			}
			bodyLines = append(bodyLines, strippedLine)
			depth = newDepth
		}
	}
	if inBlock {
		return nil, fmt.Errorf("DRIFT block %q: unclosed block (missing -- ))", currentID)
	}
	return blocks, nil
}

func buildAnticipatedDrift(block *DriftBlock) error {
	switch block.Scope {
	case DriftScopeTable:
		body := block.RawBody
		// Strip trailing comma from last non-empty line
		bodyLines := strings.Split(body, "\n")
		for i := len(bodyLines) - 1; i >= 0; i-- {
			if strings.TrimSpace(bodyLines[i]) != "" {
				bodyLines[i] = strings.TrimRight(bodyLines[i], " \t,")
				break
			}
		}
		body = strings.Join(bodyLines, "\n")
		syntheticSQL := "CREATE TABLE __drift_probe__ (\n" + body + "\n);"
		tables, _, _, err := parseDDL(syntheticSQL)
		if err != nil {
			return fmt.Errorf("buildAnticipatedDrift: parse table-scope body for block %q: %w", block.ID, err)
		}
		if len(tables) > 0 {
			block.AnticipatedTable = tables[0]
		}
	case DriftScopeFile:
		tables, indexes, _, err := parseDDL(block.RawBody)
		if err != nil {
			return fmt.Errorf("buildAnticipatedDrift: parse file-scope body for block %q: %w", block.ID, err)
		}
		if len(tables) > 0 {
			block.AnticipatedTable = tables[0]
		}
		block.AnticipatedIndexes = indexes
	}
	return nil
}

// MergeDriftGroups merges all DriftBlocks with the same ID into a single DriftGroup.
func MergeDriftGroups(allBlocks []DriftBlock) (map[string]*DriftGroup, error) {
	groups := make(map[string]*DriftGroup)
	for _, b := range allBlocks {
		g, ok := groups[b.ID]
		if !ok {
			g = &DriftGroup{ID: b.ID, Policy: b.Policy}
			groups[b.ID] = g
		}
		if g.Policy != b.Policy {
			return nil, fmt.Errorf("DRIFT group %q has conflicting policies: %s and %s", b.ID, g.Policy, b.Policy)
		}
		if b.AnticipatedTable != nil {
			g.AnticipatedColumns = append(g.AnticipatedColumns, b.AnticipatedTable.Columns...)
			g.AnticipatedUniqueKeys = append(g.AnticipatedUniqueKeys, b.AnticipatedTable.UniqueKeys...)
			g.AnticipatedForeignKeys = append(g.AnticipatedForeignKeys, b.AnticipatedTable.ForeignKeys...)
		}
		g.AnticipatedIndexes = append(g.AnticipatedIndexes, b.AnticipatedIndexes...)
	}
	return groups, nil
}

File: rust/schemify/src/drift.rs
//! Drift block parsing (ported from Go internal/schema/drift.go).

use regex::Regex;
use std::collections::HashMap;
use std::sync::OnceLock;

use crate::{
    error::{Error, Result},
    load::parse_ddl,
    schema::{DriftBlock, DriftGroup, DriftPolicy, DriftScope},
};

static DRIFT_OPEN_RE: OnceLock<Regex> = OnceLock::new();
static DRIFT_CLOSE_RE: OnceLock<Regex> = OnceLock::new();

fn drift_open_re() -> &'static Regex {
    DRIFT_OPEN_RE.get_or_init(|| {
        Regex::new(r"(?i)^\s*--\s*DRIFT\s+(\w+)\s+(DROP|DEPRECATED)\s*\(\s*$").unwrap()
    })
}

fn drift_close_re() -> &'static Regex {
    DRIFT_CLOSE_RE.get_or_init(|| Regex::new(r"^\s*--\s*\)\s*;?\s*$").unwrap())
}

pub fn strip_comment_prefix(line: &str) -> String {
    match line.find("--") {
        None => line.to_string(),
        Some(pos) => {
            let before = &line[..pos];
            let after = &line[pos + 2..];
            let stripped_after = after.strip_prefix(' ').unwrap_or(after);
            format!("{}{}", before, stripped_after)
        }
    }
}

pub fn net_paren_depth(line: &str) -> i32 {
    let mut depth: i32 = 0;
    let mut in_single = false;
    let mut in_double = false;
    for ch in line.chars() {
        match ch {
            '\'' if !in_double => {
                in_single = !in_single;
            }
            '"' if !in_single => {
                in_double = !in_double;
            }
            '(' if !in_single && !in_double => {
                depth += 1;
            }
            ')' if !in_single && !in_double => {
                depth -= 1;
            }
            _ => {}
        }
    }
    depth
}

pub fn extract_drift_blocks(
    raw_sql: &str,
    table_schema: &str,
    table_name: &str,
) -> Result<Vec<DriftBlock>> {
    let scope = if !table_schema.is_empty() || !table_name.is_empty() {
        DriftScope::Table
    } else {
        DriftScope::File
    };

    let mut blocks = Vec::new();
    let mut seen_ids: HashMap<String, DriftPolicy> = HashMap::new();

    let mut in_block = false;
    let mut current_id = String::new();
    let mut current_policy = DriftPolicy::Drop;
    let mut body_lines: Vec<String> = Vec::new();
    let mut depth: i32 = 0;

    for line in raw_sql.lines() {
        if !in_block {
            let Some(caps) = drift_open_re().captures(line) else {
                continue;
            };
            let id = caps.get(1).unwrap().as_str().to_string();
            let policy_str = caps.get(2).unwrap().as_str().to_uppercase();
            let policy = if policy_str == "DROP" {
                DriftPolicy::Drop
            } else {
                DriftPolicy::Deprecated
            };
            if let Some(existing) = seen_ids.get(&id) {
                if *existing != policy {
                    return Err(Error::LoadSchema(format!(
                        "DRIFT block {:?} has conflicting policies: {:?} and {:?}",
                        id, existing, policy
                    )));
                }
            }
            seen_ids.insert(id.clone(), policy.clone());
            in_block = true;
            current_id = id;
            current_policy = policy;
            body_lines = Vec::new();
            depth = 1;
        } else {
            let stripped = strip_comment_prefix(line);
            let is_closing = drift_close_re().is_match(line);
            let net = net_paren_depth(&stripped);
            let new_depth = depth + net;
            if new_depth == 0 {
                if is_closing {
                    blocks.push(DriftBlock {
                        id: current_id.clone(),
                        policy: current_policy.clone(),
                        raw_body: body_lines.join("\n"),
                        scope: scope.clone(),
                        table_schema: table_schema.to_string(),
                        table_name: table_name.to_string(),
                        anticipated_table: None,
                        anticipated_indexes: Vec::new(),
                    });
                    in_block = false;
                    continue;
                }
                return Err(Error::LoadSchema(format!(
                    "DRIFT block {:?}: depth reached 0 before closing -- )",
                    current_id
                )));
            }
            if new_depth < 0 {
                return Err(Error::LoadSchema(format!(
                    "DRIFT block {:?}: negative paren depth",
                    current_id
                )));
            }
            body_lines.push(stripped);
            depth = new_depth;
        }
    }

    if in_block {
        return Err(Error::LoadSchema(format!(
            "DRIFT block {:?}: unclosed block (missing -- ))",
            current_id
        )));
    }

    Ok(blocks)
}

fn trim_trailing_comma(body: &str) -> String {
    let mut owned: Vec<String> = body.lines().map(|l| l.to_string()).collect();
    for i in (0..owned.len()).rev() {
        if !owned[i].trim().is_empty() {
            let trimmed = owned[i].trim_end_matches([' ', '\t', ',']);
            owned[i] = trimmed.to_string();
            break;
        }
    }
    owned.join("\n")
}

pub fn build_anticipated_drift(block: &mut DriftBlock) -> Result<()> {
    match block.scope {
        DriftScope::Table => {
            let body = trim_trailing_comma(&block.raw_body);
            let synthetic_sql = format!("CREATE TABLE __drift_probe__ (\n{}\n);", body);
            let (tables, _, _) = parse_ddl(&synthetic_sql).map_err(|e| {
                Error::LoadSchema(format!(
                    "buildAnticipatedDrift: parse table-scope body for block {:?}: {}",
                    block.id, e
                ))
            })?;
            if !tables.is_empty() {
                block.anticipated_table = Some(tables.into_iter().next().unwrap());
            }
        }
        DriftScope::File => {
            let (tables, indexes, _) = parse_ddl(&block.raw_body).map_err(|e| {
                Error::LoadSchema(format!(
                    "buildAnticipatedDrift: parse file-scope body for block {:?}: {}",
                    block.id, e
                ))
            })?;
            if !tables.is_empty() {
                block.anticipated_table = Some(tables.into_iter().next().unwrap());
            }
            block.anticipated_indexes = indexes.into_iter().collect();
        }
    }
    Ok(())
}

pub fn merge_drift_groups(all_blocks: Vec<DriftBlock>) -> Result<HashMap<String, DriftGroup>> {
    let mut groups: HashMap<String, DriftGroup> = HashMap::new();
    for b in all_blocks {
        let id = b.id.clone();
        let g = groups.entry(id.clone()).or_insert_with(|| DriftGroup {
            id: id.clone(),
            policy: b.policy.clone(),
            ..Default::default()
        });
        if g.policy != b.policy {
            return Err(Error::LoadSchema(format!(
                "DRIFT group {:?} has conflicting policies: {:?} and {:?}",
                id, g.policy, b.policy
            )));
        }
        if let Some(t) = b.anticipated_table {
            g.anticipated_columns.extend(t.columns);
            g.anticipated_unique_keys.extend(t.unique_keys);
            g.anticipated_foreign_keys.extend(t.foreign_keys);
        }
        g.anticipated_indexes.extend(b.anticipated_indexes);
    }
    Ok(groups)
}
